### PR TITLE
SI-5892 Don't apply implicit class members in class type param annots

### PR DESCRIPTION
**Pull request by retronym**
- Abuse `Context#inSelfSuperCall` for this purpose.
- Inline `createNamer` into `completerOf(tree, typeParams)`, remove impossible case,
  and tweak inSelfSuperCall.

I originally tried this change in `Typers#typedClassDef`. Perhaps that also needs
to be addressed, but I couldn't motivate that with a failing test.

This problem can lead to cyclic inheritance errors, as was seen in Scalala.

Review by @lrytz

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Contexts.scala [hunk 1]
```diff
@@ -117,6 +117,7 @@ trait Contexts { self: Analyzer =>
     var inSelfSuperCall = false             // is this context (enclosed in) a constructor call?
     // (the call to the super or self constructor in the first line of a constructor)
     // in this context the object's fields should not be in scope
+    // This mode is also used when typing class type parameters.
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Contexts.scala:121]: do you think the above comment (not the part you wrote, but the line above) is correct? it seems that `inSelfSuperCall` is only used in `Context#implicitss`, and it has nothing to do with the visibility of fields.

File: src/compiler/scala/tools/nsc/typechecker/Namers.scala [hunk 2]
```diff
@@ -540,8 +540,17 @@ trait Namers extends MethodSynthesis {
         //@M! TypeDef's type params are handled differently
         //@M e.g., in [A[x <: B], B], A and B are entered first as both are in scope in the definition of x
         //@M x is only in scope in `A[x <: B]'
-        if (!tree.symbol.isAbstractType) //@M TODO: change to isTypeMember ?
-          createNamer(tree) enterSyms tparams
+        if (!tree.symbol.isAbstractType) { //@M TODO: change to isTypeMember ?
+          val scope = tree match {
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Namers.scala:544]: seems you could write more concisely

```
val scope = context.makeNewScope(tree, tree.symbol)
tree match {
  case _: ClassDef => scope. inSelfSuperCall = true
  case _ =>
}
```

also, a comment would be helpful, future readers will have a hard time figuring out the intent.

File: src/compiler/scala/tools/nsc/typechecker/Contexts.scala (hunk 1 above)
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Contexts.scala:121]: I must admit I didn't reverse engineer the other usages this flag fully. Perhaps that comment refers to the fact that self/super constructor arguments can't be adapated by an implicit class member? I'll dig into it a bit more and document my findings.

In general, the different handling of the in-scope implicits and the regular scope give me the creeps, though. (See this comment for another example: https://github.com/scala/scala/pull/697/files#L1R648)

**Comment by adriaanm** [comment]
what's the status of this pullreq -- could the reviewer give an explicit thumbs up/down?

**Comment by retronym** [comment]
The submitter will give it the downward pointed thumb -- I think we can find a more principled fix, that also considers method type parameter annotations and non implicit calls.

What should be in scope in a method type parameter declaration? Concretely, should this be an error? 

``` scala
scala> class a(a: Any) extends annotation.StaticAnnotation
defined class a

scala> class C { def id(a: Int) = a; def foo[@a(id(0)) _] = () }
defined class C
```

**Pull request closed without merging**